diff --git a/cmake.deps/cmake/LibvtermCMakeLists.txt b/cmake.deps/cmake/LibvtermCMakeLists.txt
index c197523786..5fe2e86574 100644
--- a/cmake.deps/cmake/LibvtermCMakeLists.txt
+++ b/cmake.deps/cmake/LibvtermCMakeLists.txt
@@ -15,7 +15,7 @@ install(TARGETS vterm ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
 install(FILES include/vterm.h include/vterm_keycodes.h
   DESTINATION include)
 
-if(NOT WIN32)
+if(NOT WIN32 AND NOT ZOS)
   file(GLOB BIN_SOURCES ${CMAKE_SOURCE_DIR}/bin/*.c)
   foreach(EXE_C ${BIN_SOURCES})
     get_filename_component(target_name ${EXE_C} NAME_WE)
