diff --git a/src/nvim/bufwrite.c b/src/nvim/bufwrite.c
index 27de03954a..35c0c754fc 100644
--- a/src/nvim/bufwrite.c
+++ b/src/nvim/bufwrite.c
@@ -756,7 +756,7 @@ static int buf_write_make_backup(char *fname, bool append, FileInfo *file_info_o
         }
       }
       int fd = os_open(IObuff,
-                       O_CREAT|O_WRONLY|O_EXCL|O_NOFOLLOW, perm);
+                       O_CREAT|O_WRONLY|O_EXCL, perm);
       if (fd < 0) {           // can't write in directory
         *backup_copyp = true;
       } else {
@@ -1477,6 +1477,20 @@ restore_backup:
         }
         goto fail;
       }
+
+#ifdef __MVS__
+      if (newfile == true)
+        __tag_new_file(fd);
+      else if (file_info_old.stat.st_tag.ft_ccsid == FT_UNTAGGED) {
+        __setfdccsid(fd, 0);
+        struct f_cnvrt cvtreq = {SETCVTON, 0, 1047};
+        if (curbuf->needs_conversion)
+          fcntl(fd, F_CONTROL_CVT, &cvtreq);
+      }
+      else
+        __chgfdccsid(fd, file_info_old.stat.st_tag.ft_ccsid);
+#endif
+
       write_info.bw_fd = fd;
     }
     err = set_err(NULL);
@@ -1847,7 +1861,7 @@ restore_backup:
 
       if (org == NULL
           || (empty_fd = os_open(org,
-                                 O_CREAT | O_EXCL | O_NOFOLLOW,
+                                 O_CREAT | O_EXCL ,
                                  perm < 0 ? 0666 : (perm & 0777))) < 0) {
         emsg(_(e_patchmode_cant_touch_empty_original_file));
       } else {
