diff --git a/src/nvim/memfile.c b/src/nvim/memfile.c
index a1713edb66..27f8292a11 100644
--- a/src/nvim/memfile.c
+++ b/src/nvim/memfile.c
@@ -750,7 +750,7 @@ static bool mf_do_open(memfile_T *mfp, char *fname, int flags)
     emsg(_("E300: Swap file already exists (symlink attack?)"));
   } else {
     // try to open the file
-    mfp->mf_fd = os_open(mfp->mf_fname, flags | O_NOFOLLOW, S_IREAD | S_IWRITE);
+    mfp->mf_fd = os_open(mfp->mf_fname, flags, S_IREAD | S_IWRITE);
   }
 
   // If the file cannot be opened, use memory only
