diff --git a/cmake.deps/cmake/BuildTreesitter.cmake b/cmake.deps/cmake/BuildTreesitter.cmake
index 51aa40e763..6e4cab03e1 100644
--- a/cmake.deps/cmake/BuildTreesitter.cmake
+++ b/cmake.deps/cmake/BuildTreesitter.cmake
@@ -5,6 +5,7 @@ endif()
 get_externalproject_options(treesitter ${DEPS_IGNORE_SHA})
 ExternalProject_Add(treesitter
   DOWNLOAD_DIR ${DEPS_DOWNLOAD_DIR}/treesitter
+  PATCH_COMMAND patch -p1 < ${CMAKE_SOURCE_DIR}/../../deps_patches/treesitter.patch
   CMAKE_ARGS ${DEPS_CMAKE_ARGS} ${TREESITTER_ARGS}
   ${EXTERNALPROJECT_OPTIONS})

