diff --git a/src/nvim/os/pty_proc_unix.c b/src/nvim/os/pty_proc_unix.c
index 10a464bbc3..4be9c0b2f2 100644
--- a/src/nvim/os/pty_proc_unix.c
+++ b/src/nvim/os/pty_proc_unix.c
@@ -15,6 +15,8 @@
 # include <libutil.h>
 #elif defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
 # include <util.h>
+#elif defined(__MVS__)
+# include <fcntl.h>
 #elif defined(__sun)
 # include <fcntl.h>
 # include <signal.h>
@@ -338,14 +340,18 @@ static void init_termios(struct termios *termios) FUNC_ATTR_NONNULL_ALL
   termios->c_cc[VERASE] = 0x7f;
   termios->c_cc[VKILL] = 0x1f & 'U';
   termios->c_cc[VEOF] = 0x1f & 'D';
+#ifndef __MVS__
   termios->c_cc[VEOL] = _POSIX_VDISABLE;
   termios->c_cc[VEOL2] = _POSIX_VDISABLE;
+#endif
   termios->c_cc[VSTART] = 0x1f & 'Q';
   termios->c_cc[VSTOP] = 0x1f & 'S';
   termios->c_cc[VSUSP] = 0x1f & 'Z';
+#ifndef __MVS__
   termios->c_cc[VREPRINT] = 0x1f & 'R';
   termios->c_cc[VWERASE] = 0x1f & 'W';
   termios->c_cc[VLNEXT] = 0x1f & 'V';
+#endif
   termios->c_cc[VMIN] = 1;
   termios->c_cc[VTIME] = 0;
 }
