diff --git a/src/nvim/fileio.c b/src/nvim/fileio.c
index 6aa68fe..1d84a9f 100644
--- a/src/nvim/fileio.c
+++ b/src/nvim/fileio.c
@@ -86,6 +86,10 @@
 # include "nvim/charset.h"
 #endif
 
+#ifdef __MVS__
+#include <zos.h>
+#endif
+
 // For compatibility with libuv < 1.20.0 (tested on 1.18.0)
 #ifndef UV_FS_COPYFILE_FICLONE
 # define UV_FS_COPYFILE_FICLONE 0
@@ -440,6 +444,10 @@ int readfile(char *fname, char *sfname, linenr_T from, linenr_T lines_to_skip,
     fd = os_open(fname, O_RDONLY, 0);
   }
 
+#ifdef __MVS__
+  curbuf->needs_conversion = __file_needs_conversion(fd);
+#endif
+
   if (fd < 0) {                     // cannot open at all
     msg_scroll = msg_save;
     if (!newfile) {
