diff --git a/src/nvim/undo.c b/src/nvim/undo.c
index 8d791a4c38..01c67286ae 100644
--- a/src/nvim/undo.c
+++ b/src/nvim/undo.c
@@ -1249,7 +1249,7 @@ void u_write_undo(const char *const name, const bool forceit, buf_T *const buf,
     goto theend;
   }
 
-  int fd = os_open(file_name, O_CREAT|O_WRONLY|O_EXCL|O_NOFOLLOW, perm);
+  int fd = os_open(file_name, O_CREAT|O_WRONLY|O_EXCL, perm);
   if (fd < 0) {
     semsg(_(e_not_open), file_name);
     goto theend;
