diff --git a/src/nvim/CMakeLists.txt b/src/nvim/CMakeLists.txt
index b8160fea4b..f229292592 100644
--- a/src/nvim/CMakeLists.txt
+++ b/src/nvim/CMakeLists.txt
@@ -144,10 +144,6 @@ elseif(CMAKE_C_COMPILER_ID MATCHES "Clang")
     target_compile_options(main_lib INTERFACE -Wno-c11-extensions)
   endif()

-  # workaround for clang-11 on macOS, supported on later versions
-  if(NOT APPLE AND NOT CMAKE_CROSSCOMPILING)
-    target_link_libraries(nvim_bin PRIVATE -Wl,--no-undefined)
-  endif()
 endif()

 # Platform specific options
