diff --git a/src/nvim/os/env.c b/src/nvim/os/env.c
index 940d223..e502413 100644
--- a/src/nvim/os/env.c
+++ b/src/nvim/os/env.c
@@ -207,7 +207,7 @@ size_t os_get_fullenv_size(void)
 #else
 # if defined(HAVE__NSGETENVIRON)
   char **environ = *_NSGetEnviron();
-# else
+# elif !defined(environ)
   extern char **environ;
 # endif
 
@@ -265,7 +265,7 @@ void os_copy_fullenv(char **env, size_t env_size)
 #else
 # if defined(HAVE__NSGETENVIRON)
   char **environ = *_NSGetEnviron();
-# else
+# elif !defined(environ)
   extern char **environ;
 # endif
 
@@ -324,7 +324,7 @@ char *os_getenvname_at_index(size_t index)
 #else
 # if defined(HAVE__NSGETENVIRON)
   char **environ = *_NSGetEnviron();
-# else
+# elif !defined(environ)
   extern char **environ;
 # endif
 
